src/psc_datalogger/gui.py

```diff
@@ -0,0 +1,227 @@
+
+    def __init__(self, callback: Callable[[str], None]) -> None:
+        super().__init__()
+
+        self.create_widgets()
+
+        self.filename_callback = callback
+
+        self.filename: Optional[str] = None
+
+    def create_widgets(self) -> None:
+        file_layout = QHBoxLayout()
+        # This is actually a pop-up dialog, not an inline widget
+        self.dialog = QFileDialog(None, "Select output logfile")
+        self.dialog.setFileMode(QFileDialog.FileMode.AnyFile)
+        self.dialog.setAcceptMode(QFileDialog.AcceptMode.AcceptSave)
+        self.dialog.setNameFilter("CSV (*.csv)")
+        self.dialog.setDefaultSuffix(".csv")
+        self.dialog.fileSelected.connect(self.file_selected)
+
+        file_button = QPushButton("Select Logfile...")
+        # Suppress error that `exec` doesn't technically match signature of a PYQT_SLOT
+        # - it doesn't affect anything, the return value is discarded
+        file_button.clicked.connect(self.dialog.exec)  # type: ignore
+        file_layout.addWidget(file_button)
+
+        self.file_readback = QLineEdit()
+        self.file_readback.setDisabled(True)
+        file_layout.addWidget(self.file_readback)
+
+        self.setLayout(file_layout)
+
+    def file_selected(self, new_path: str) -> None:
+        """Handle a new filepath being set"""
+        self.filename = new_path
+        self.file_readback.setText(new_path)
+        self.filename_callback(new_path)
+
+
+class AgilentWidgets(QGroupBox):
+    """Contains widgets describing a single Agilent 3458A instrument"""
+
+    def __init__(
+        self, instrument_number: int, checked: bool, instrument_changed: Callable
+    ) -> None:
+        """
+        Args:
+          instrument_number: The number to use to identify this instance
+          checked: True if the widgets should be enabled by default
+          instrument_changed: Callback to be called whenever an address changes
+        """
+        super().__init__(f"Instrument {instrument_number}")
+        self.instrument_number = instrument_number
+
+        self.setCheckable(True)
+        self.setChecked(checked)
+        self.toggled.connect(instrument_changed)
+
+        self.create_widgets(instrument_changed)
+
+    def create_widgets(self, instrument_changed: Callable) -> None:
+        address_label = QLabel("GPIB Address")
+        self.address_input_box = QLineEdit()
+        int_only = QIntValidator()
+        int_only.setRange(0, 30)
+        self.address_input_box.setValidator(int_only)
+        self.address_input_box.editingFinished.connect(instrument_changed)
+        self.voltage_radiobutton = QRadioButton("Voltage")
+        self.voltage_radiobutton.setChecked(True)
+        self.voltage_radiobutton.toggled.connect(instrument_changed)
+        self.temperature_radiobutton = QRadioButton("Temperature")
+        # Don't need to .connect() the second button as changing the first one
+        # triggers a callback that will check both buttons
```

Comment:
The "voltage" and "temperature" radio buttons are, by black magic of Qt, bound together into a hidden group, such that only one of the two can be pressed at once. Hence, changing either button will effectively change both, as both will change state.

I think this would break if there were three or more buttons, but as-is it works.